### PR TITLE
quick fix openm install and 8.2 version

### LINKED ISSUE #1144: Updating OpenMM module to fit v8.2.0
## **Desired feature/enhancement**

Update openmm module to fit new v8.2.0 openmm implementation and python api

## **Motivation**

Maintain the module functional

### LINKED ISSUE #1153: OpenMM Module- "Exception in task execution: 'NoneType' object is not iterable"
Fristly thanks for the great program and easy installation process.

## **OpenMM module**
Running Openmm module to refine top models:

Here is my full
```toml
run_dir = "haddock_files"
mode = "local"
ncores = 31
self_contained = true
postprocess = true
clean = true
molecules =  [
    "clean_ab.pdb",
    "clean_ag.pdb"
    ]

[topoaa]

[rigidbody]
sampling = 2000
randremoval = true
tolerance = 5
ambig_fname = "ambig.tbl"
unambig_fname = "ab_unambig.tbl"

[clustfcc]

[seletopclusts]
top_cluster = 10
top_models = 20

[flexref]
ambig_fname = "ambig.tbl"
unambig_fname = "ab_unambig.tbl"
randremoval = true

[emref]
ambig_fname = "ambig.tbl"
unambig_fname = "ab_unambig.tbl"
randremoval = true

[mdref]
ambig_fname = "ambig.tbl"
unambig_fname = "ab_unambig.tbl"
randremoval = true

[clustfcc]

#Contact map and alanine scanning
[seletopclusts]
top_cluster = 10
top_models = 10

[caprieval]
reference_fname = "5xmh_ref.pdb"

[seletopclusts]
top_cluster = 3
top_models = 3

[openmm]
forcefield = 'amber14/protein.ff14SB.xml'
explicit_solvent_model = 'amber14/tip3p.xml'
timestep_ps = 0.002
constraints = 'HBonds'
equilibration_timesteps = 1000
simulation_timesteps = 1000
```

## **Expected behavior**
Expecting refined model output

## **Actual behavior**
Run fine until step 11:

```text
[<time>__init__ INFO] [seletopclusts] took 0 seconds
[<time>__init__ INFO] Running [openmm] module
[<time>__init__ INFO] [openmm] Creating directory pdbfixer
[<time>__init__ INFO] [openmm] Creating directory solvation_boxes
[<time>__init__ INFO] [openmm] Creating directory intermediates
[<time>__init__ INFO] [openmm] Creating directory md_raw_output
[<time>__init__ INFO] [openmm] Creating directory openmm_output
[<time>__init__ INFO] [openmm] Creating directory simulation_stats
[<time>libutil INFO] Selected 9 cores to process 9 jobs, with 32 maximum available cores.
[<time>libparallel INFO] Using 9 cores
[<time>openmm INFO] Fixing pdb: cluster_1_model_1.pdb (path ../10_seletopclusts/cluster_1_model_1.pdb)
[<time>openmm INFO] Fixing pdb: cluster_1_model_2.pdb (path ../10_seletopclusts/cluster_1_model_2.pdb)
[<time>openmm INFO] Fixing pdb: cluster_1_model_3.pdb (path ../10_seletopclusts/cluster_1_model_3.pdb)
[<time>openmm INFO] Fixing pdb: cluster_2_model_1.pdb (path ../10_seletopclusts/cluster_2_model_1.pdb)
[<time>openmm INFO] Fixing pdb: cluster_2_model_2.pdb (path ../10_seletopclusts/cluster_2_model_2.pdb)
[<time>openmm INFO] Fixing pdb: cluster_2_model_3.pdb (path ../10_seletopclusts/cluster_2_model_3.pdb)
[<time>openmm INFO] Fixing pdb: cluster_3_model_1.pdb (path ../10_seletopclusts/cluster_3_model_1.pdb)
[<time>openmm INFO] Fixing pdb: cluster_3_model_2.pdb (path ../10_seletopclusts/cluster_3_model_2.pdb)
[<time>openmm INFO] Fixing pdb: cluster_3_model_3.pdb (path ../10_seletopclusts/cluster_3_model_3.pdb)
[<time>openmm INFO] Building solvation box for: cluster_1_model_3.pdb
[<time>openmm INFO] Building solvation box for: cluster_1_model_2.pdb
[<time>openmm INFO] Building solvation box for: cluster_2_model_1.pdb
[<time>openmm INFO] Building solvation box for: cluster_1_model_1.pdb
[<time>openmm INFO] Building solvation box for: cluster_2_model_3.pdb
[<time>openmm INFO] Building solvation box for: cluster_3_model_2.pdb
[<time>openmm INFO] Building solvation box for: cluster_3_model_3.pdb
[<time>openmm INFO] Building solvation box for: cluster_3_model_1.pdb
[<time>openmm INFO] Building solvation box for: cluster_2_model_2.pdb
[<time>openmm INFO] Solvating and adding ions to the system:padding 1.0 nm, ions conc. : 0.15 M
[<time>openmm INFO] Solvating and adding ions to the system:padding 1.0 nm, ions conc. : 0.15 M
[<time>openmm INFO] Solvating and adding ions to the system:padding 1.0 nm, ions conc. : 0.15 M
[<time>openmm INFO] Solvating and adding ions to the system:padding 1.0 nm, ions conc. : 0.15 M
[<time>openmm INFO] Solvating and adding ions to the system:padding 1.0 nm, ions conc. : 0.15 M
[<time>openmm INFO] Solvating and adding ions to the system:padding 1.0 nm, ions conc. : 0.15 M
[<time>openmm INFO] Solvating and adding ions to the system:padding 1.0 nm, ions conc. : 0.15 M
[<time>openmm INFO] Solvating and adding ions to the system:padding 1.0 nm, ions conc. : 0.15 M
[<time>openmm INFO] Solvating and adding ions to the system:padding 1.0 nm, ions conc. : 0.15 M
[<time>openmm INFO] Equilibrating solvation box from solvation_boxes/cluster_1_model_2.pdb...
[<time>openmm INFO] Equilibrating solvation box from solvation_boxes/cluster_1_model_3.pdb...
[<time>openmm INFO] Equilibrating solvation box from solvation_boxes/cluster_1_model_1.pdb...
[<time>openmm INFO] Equilibrating solvation box from solvation_boxes/cluster_3_model_3.pdb...
[<time>openmm INFO] Equilibrating solvation box from solvation_boxes/cluster_2_model_1.pdb...
[<time>openmm INFO] Equilibrating solvation box from solvation_boxes/cluster_3_model_1.pdb...
[<time>openmm INFO] Equilibrating solvation box from solvation_boxes/cluster_2_model_2.pdb...
[<time>openmm INFO] Equilibrating solvation box from solvation_boxes/cluster_3_model_2.pdb...
[<time>openmm INFO] Equilibrating solvation box from solvation_boxes/cluster_2_model_3.pdb...
[<time>openmm INFO] Adding protein constraints
[<time>openmm INFO] Adding protein constraints
[<time>openmm INFO] Adding protein constraints
[<time>openmm INFO] Adding protein constraints
[<time>openmm INFO] Adding protein constraints
[<time>openmm INFO] Adding protein constraints
[<time>openmm INFO] Adding protein constraints
[<time>openmm INFO] Adding protein constraints
[<time>openmm INFO] Adding protein constraints
[<time>openmm INFO] Minimizing system energy
[<time>openmm INFO] starting explicit solvent openMM simulation with file: None
[<time>libparallel WARNING] Exception in task execution: 'NoneType' object is not iterable
[<time>openmm INFO] Minimizing system energy
[<time>openmm INFO] starting explicit solvent openMM simulation with file: None
[<time>libparallel WARNING] Exception in task execution: 'NoneType' object is not iterable
[<time>openmm INFO] Minimizing system energy
[<time>openmm INFO] starting explicit solvent openMM simulation with file: None
[<time>libparallel WARNING] Exception in task execution: 'NoneType' object is not iterable
[<time>openmm INFO] Minimizing system energy
[<time>openmm INFO] starting explicit solvent openMM simulation with file: None
[<time>libparallel WARNING] Exception in task execution: 'NoneType' object is not iterable
[<time>openmm INFO] Minimizing system energy
[<time>openmm INFO] starting explicit solvent openMM simulation with file: None
[<time>libparallel WARNING] Exception in task execution: 'NoneType' object is not iterable
[<time>openmm INFO] Minimizing system energy
[<time>openmm INFO] Minimizing system energy
[<time>openmm INFO] starting explicit solvent openMM simulation with file: None
[<time>libparallel WARNING] Exception in task execution: 'NoneType' object is not iterable
[<time>openmm INFO] starting explicit solvent openMM simulation with file: None
[<time>libparallel WARNING] Exception in task execution: 'NoneType' object is not iterable
[<time>openmm INFO] Minimizing system energy
[<time>openmm INFO] starting explicit solvent openMM simulation with file: None
[<time>libparallel WARNING] Exception in task execution: 'NoneType' object is not iterable
[<time>openmm INFO] Minimizing system energy
[<time>openmm INFO] starting explicit solvent openMM simulation with file: None
[<time>libparallel WARNING] Exception in task execution: 'NoneType' object is not iterable
[<time>libparallel INFO] 9 tasks finished
[<time>libutil ERROR] No output models generated. Check Openmm Execution and logfile.
Traceback (most recent call last):
  File "~miniconda3/envs/haddock_pip/lib/python3.9/site-packages/haddock/libs/libutil.py", line 335, in log_error_and_exit
    yield
  File "~miniconda3/envs/haddock_pip/lib/python3.9/site-packages/haddock/clis/cli.py", line 195, in main
    workflow.run()
  File "~miniconda3/envs/haddock_pip/lib/python3.9/site-packages/haddock/libs/libworkflow.py", line 43, in run
    step.execute()
  File "~miniconda3/envs/haddock_pip/lib/python3.9/site-packages/haddock/libs/libworkflow.py", line 162, in execute
    self.module.run()  # type: ignore
  File "~miniconda3/envs/haddock_pip/lib/python3.9/site-packages/haddock/modules/__init__.py", line 245, in run
    self._run()
  File "~miniconda3/envs/haddock_pip/lib/python3.9/site-packages/haddock/modules/refinement/openmm/__init__.py", line 164, in _run
    self.finish_with_error(
  File "~miniconda3/envs/haddock_pip/lib/python3.9/site-packages/haddock/modules/__init__.py", line 317, in finish_with_error
    raise RuntimeError(reason)
RuntimeError: No output models generated. Check Openmm Execution and logfile.
```

## **Steps to reproduce the behavior**
Docking two molecules, antibody and antigen with openmm refinement. There is a gap in numbering due to two chains of antibody being merged and restrained ie 1...120...1001-1120. (?cause of error)

## **Suggestions on how to fix it**
I presume this is an issue with  PDB numbering, but the chain linking is per the haddock tutorial. Should there be a way to renumber and rechain the protein e.g. with pdb_tools, before running openmm?

## **Version**
HADDOCK3 v2024.10.0b6 on 2024-11-19 11:06:00

### PR DESCRIPTION
**Pull request by VGPReys**
## Checklist

- [x] Tests added for the new code
- [ ] Documentation added for the code changes
- [x] Does not break licensing
- [x] Does not add any dependencies, if it does please add a thorough explanation

## Summary of the Pull Request  

- Update OpenMM installation instructions
- Small modification of the code to make sure implicit solvent still functional
- Adding custom solution for chain break detection: It has been observed that chain breaks in the OpenMM module are making the simulation crash. Therefore we are now pre-processing the input files in the search of such breaks and adding TER statements between them, allowing to run the simulation without issues.

## Related Issue

Closes #1144
Closes #1153

### CODE REVIEW

File: docs/INSTALL.md [hunk 1]
```diff
@@ -90,16 +90,23 @@ information.
 ## `openmm`
 
 1. Install the latest version of openmm and pdbfixer
+
+The current implementation requires both OpanMM and pdbfixer, which should be installed using conda.
```
Comment (amjjbonvin) [inline docs/INSTALL.md:94]: Typo: OpanMM -> OpenMM

File: src/haddock/modules/refinement/openmm/__init__.py
Comment (amjjbonvin) [inline src/haddock/modules/refinement/openmm/__init__.py:1]: Does it mean that HADDOCK3 must be installed with conda per se?  Or is a venv installation also possible?

File: src/haddock/modules/refinement/openmm/openmm.py [hunk 2]
```diff
@@ -716,11 +716,17 @@ def run_openmm(
...
+            dof = statereporter._dof
+        except AttributeError:
+            # may not exists if in implicit solvent..
+            dof = system.getNumParticles() * 3
         # Makes sure temperature of the system is reached
         self._stabilize_temperature(
             simulation,
-            qtemp,
-            statereporter._dof,
+            self.params["temperature_kelvin"],
```
Comment (mgiulini) [inline src/haddock/modules/refinement/openmm/openmm.py:728]: so before the code was always passing the same temperature within the loop???

**Review by mgiulini — APPROVED** [review]

File: src/haddock/modules/refinement/openmm/openmm.py (hunk 2 above)
Comment (VGPReys) [reply src/haddock/modules/refinement/openmm/openmm.py:728]: It is still doing so, I just changed the entry point of the temperature.
Not it is the one that was been set by the user,
Before it was the one set by the user divided by 10, multiplied by 10

Comment (mgiulini) [reply src/haddock/modules/refinement/openmm/openmm.py:728]: I see now, thanks!

File: src/haddock/modules/refinement/openmm/__init__.py [hunk 3]
```diff
@@ -98,26 +98,17 @@ def confirm_installation(cls) -> None:
...
-        checkOpenMM = run_subprocess("conda list openmm --json")
-        checkPdbfixer = run_subprocess("conda list pdbfixer --json")
-
-        if (checkOpenMM == "[]"):
+        try:
+            import openmm
+        except ModuleNotFoundError:
             raise ThirdPartyIntallationError(
-                "OpenMM is not installed in conda."
+                "OpenMM pdbfixer is not installed."
```
Comment (VGPReys) [inline src/haddock/modules/refinement/openmm/__init__.py:105]:
```suggestion
                "OpenMM is not installed."

```

[the remark this answers is not in the document]
Comment (VGPReys) [reply src/haddock/modules/refinement/openmm/__init__.py]: Copy-paste issue

[the remark this answers is not in the document]
Comment (VGPReys) [reply src/haddock/modules/refinement/openmm/__init__.py]: Solved

File: docs/INSTALL.md [hunk 4]
```diff
@@ -89,17 +89,37 @@ information.
...
 ## `openmm`
 
-1. Install the latest version of openmm and pdbfixer
+Install the latest version of OpenMM and pdbfixer
+
+### In venv
+
+```bash
+# Create a new virtual env
+python3.11 -m venv hd3-openmm
```
[the remark this answers is not in the document]
Comment (VGPReys) [reply docs/INSTALL.md:98]: Yes maybe, I still have to make sure it is functional for all python3 version...

[the remark this answers is not in the document]
Comment (VGPReys) [reply docs/INSTALL.md:98]: Not functional for python3.12 ...

**Comment by amjjbonvin** [comment]
So we need to specify the working Python versions to install haddock3 with openmm.

**Comment by VGPReys** [comment]
More complicated than what I thought...
conda-forge and pip do not behave the same.
deposited versions are not the same.
Something functional with conda is not with venv and vice-versa...
I am about to have two different installation instructions, depending on how haddock3 is installed.
Also they are constantly providing updates on their side...

**Comment by rvhonorato** [comment]
> More complicated than what I thought... conda-forge and pip do not behave the same. deposited versions are not the same. Something functional with conda is not with venv and vice-versa... I am about to have two different installation instructions, depending on how haddock3 is installed. Also they are constantly providing updates on their side...

- https://anaconda.org/conda-forge/openmm/files?version=8.2.0
- https://pypi.org/project/OpenMM/8.2.0/

It's 8.2 on both of them

**Comment by rvhonorato** [comment]
I know you like to do things on trial-and-error instead of reading the documentation first, so just in case you have not yet spent some time trying to figure out with which version openmm 8.2 is compatible, you can find this information either in the package info on conda-forge (https://anaconda.org/conda-forge/openmm/files/modal/info/672eb96e41d1a787229656f8) or on pypi

File: docs/INSTALL.md [hunk 5]
```diff
@@ -89,17 +89,41 @@ information.
...
+The venv installation instruction uses `pip` and requires a python version >=3.10 (not functional with python3.9).
+
+```bash
+# Create a new virtual env
+python3.12 -m venv hd3-openmm
+source hd3-openmm/bin/activate
+# Install haddock3
+pip install .
+# Install OpenMM
+pip install openmm
```
Comment (VGPReys) [inline docs/INSTALL.md:105]:
```suggestion
pip install openmm==8.2.0
```

File: docs/INSTALL.md [hunk 6]
```diff
@@ -89,17 +89,41 @@ information.
...
 # Activate the haddock3 env
-conda activate haddock3
+conda activate hd3-p39-openmm
+# Install haddock3
+pip install haddock3
 # libstdxx must be installed, maybe already present in your system
 conda install -c conda-forge libstdcxx-ng
 # Install OpenMM and related tools/binaries
-conda install -c conda-forge openmm pdbfixer
+conda install -c conda-forge openmm==8.2.0 pdbfixer
```
Comment (VGPReys) [inline docs/INSTALL.md:123]:
```suggestion
conda install -c conda-forge openmm==8.2.0 pdbfixer==1.10
```

**Comment by VGPReys** [comment]
Indeed good idea to specify the version/tag.

But I still do not understand why under conda I manage to install it with python3.9 and not for venv pip...

**Review by mgiulini — APPROVED** [review]

**Pull request merged**